gl3n/aabb.d

```diff
@@ -1,204 +1,306 @@
 module gl3n.aabb;
 
 private {
-    import gl3n.linalg : Vector, vec3;
+    import gl3n.linalg : Vector;
     import gl3n.math : almost_equal;
 }
 
 
 /// Base template for all AABB-types.
 /// Params:
 /// type = all values get stored as this type
-struct AABBT(type) {
+struct AABBT(type, uint dimension_ = 3) {
     alias type at; /// Holds the internal type of the AABB.
-    alias Vector!(at, 3) vec3; /// Convenience alias to the corresponding vector type.
+    alias Vector!(at, dimension_) vec; /// Convenience alias to the corresponding vector type.
+    alias dimension = dimension_;
+    static assert(dimension_ > 0, "0 dimensional AABB don't exist.");
 
-    vec3 min = vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0); /// The minimum of the AABB (e.g. vec3(0, 0, 0)).
-    vec3 max = vec3(cast(at)0.0, cast(at)0.0, cast(at)0.0); /// The maximum of the AABB (e.g. vec3(1, 1, 1)).
+    vec min = vec(cast(at)0.0); /// The minimum of the AABB (e.g. vec(0, 0, 0)).
+    vec max = vec(cast(at)0.0); /// The maximum of the AABB (e.g. vec(1, 1, 1)).
 
-    @safe pure nothrow:
+    //@safe pure nothrow://TODO UNCOMMENT
 
     /// Constructs the AABB.
     /// Params:
     /// min = minimum of the AABB
     /// max = maximum of the AABB
-    this(vec3 min, vec3 max) {
+    this(vec min, vec max) {
         this.min = min;
         this.max = max;
     }
 
     /// Constructs the AABB around N points (all points will be part of the AABB).
-    static AABBT from_points(vec3[] points) {
+    static AABBT from_points(vec[] points) {
         AABBT res;
 
-        if(points.length == 0) {
+        if (points.length == 0) {
             return res;
         }
 
         res.min = points[0];
         res.max = points[0];
-        foreach(v; points[1..$]) {
+        foreach (v; points[1..$]) {
             res.expand(v);
         }
-        
+
         return res;
     }
 
-    unittest {
-        AABBT!at a = AABBT!at(vec3(cast(at)0.0, cast(at)1.0, cast(at)2.0), vec3(cast(at)1.0, cast(at)2.0, cast(at)3.0));
-        assert(a.min == vec3(cast(at)0.0, cast(at)1.0, cast(at)2.0));
-        assert(a.max == vec3(cast(at)1.0, cast(at)2.0, cast(at)3.0));
+    // Convenience function to get a dimension_ sized vector for unittests
+    version(unittest)
+    private static vec sizedVec(T)(T[] values){
```

Comment:
This is used only for unittests, and allows to easily create a `dimension_` vector of type `at`